peers/cotopicpool.go

```diff
@@ -0,0 +1,80 @@
+package peers
+
+import (
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/p2p/discover"
+	"github.com/status-im/status-go/params"
+	"github.com/status-im/status-go/signal"
+)
+
+// MailServerDiscoveryTopic topic name for mailserver discovery.
+const MailServerDiscoveryTopic = "mailserver"
+
+// MailServerDiscoveryLimits default mailserver discovery limits.
+var MailServerDiscoveryLimits = params.Limits{Min: 3, Max: 3}
+
+// newCacheOnlyTopicPool returns instance of CacheOnlyTopicPool.
+func newCacheOnlyTopicPool(t *TopicPool) *cacheOnlyTopicPool {
+	return &cacheOnlyTopicPool{
+		TopicPool: t,
+	}
+}
+
+// cacheOnlyTopicPool handles a mail server topic pool.
+type cacheOnlyTopicPool struct {
+	*TopicPool
+}
+
+// MaxReached checks if the max allowed peers is reached or not. When true
+// peerpool will stop the discovery process on this TopicPool.
+// Main difference with basic TopicPool is we want to stop discovery process
+// when the number of cached peers eq/exceeds the max limit.
+func (t *cacheOnlyTopicPool) MaxReached() bool {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+	if t.limits.Max == 0 {
+		return true
+	}
+	peers := t.cache.GetPeersRange(t.topic, t.limits.Max)
+	return len(peers) >= t.limits.Max
+}
+
+var sendEnodeDiscovered = signal.SendEnodeDiscovered
+
+// ConfirmAdded calls base TopicPool ConfirmAdded method and sends a signal
+// confirming the enode has been discovered.
+func (t *cacheOnlyTopicPool) ConfirmAdded(server *p2p.Server, nodeID discover.NodeID) {
+	t.TopicPool.ConfirmAdded(server, nodeID)
+	sendEnodeDiscovered(nodeID.String(), string(t.topic))
+}
+
+// StopSearch stops the search for current topic pool, in addition to TopicPool
+// StopSearch here we are resetting topic pool limits to prevent discover to
+// be restarted, and we are removing all connected peers as we don't want them
+// to be occupying mailserver slots.
```

Comment:
i dont get the reasoning behind this. why does discovery need to remove connected peers?

Reply:
There are two sources of mail servers (Discovery + user's mail servers). Also, user can select arbitrary mail server from the list. In order to keep the logic on status-react the same, the idea in this PR is to just discover mail server peers, pass them to status-react in order to update the list and follow the existing logic.

The other solution can be not to remove mail server peers and just send them to status-react (I would then reduce the limits to 1 or 2). If user selects a different one though, we would end up with at least two mail server connections unless we implement additional logic. That's also fine at this point.

Reply:
@adambabik i didnt get it. this topic doesn't enforce connections, right? it just sends signals when something is discovered. as you say status-react accumulates those mail servers and then selects one, from discovered or hardcoded. when 3 of them are discovered - we stop the pool and disconnect from any mail server that was selected from discovered since discovery was started. so, why we will end up with 2 connections?

Reply:
> this topic doesn't enforce connections, right?

What do you mean by `enforce connections`? Peers found using this topic will be added to the server peers and a signal will be sent only after receiving a confirmation that a peer got connected. Otherwise, it's not possible to figure out if a found mail server is alive.

> so, why we will end up with 2 connections if we wont disconnect from them?

If we don't disconnect found mail server peers using Disc V5 and user will select a different one, then we will have connections to discovered mail server + selected one, so at least two nodes provided that the min limit is set to 1.

Reply:
> What do you mean by enforce connections? Peers found using this topic will be added to the server peers and a signal will be sent only after receiving a confirmation that a peer got connected. Otherwise, it's not possible to figure out if a found mail server is alive.

now it makes sense why peers are getting dropped. my understanding was that a client will validate those nodes and then select. i though that this was implemented like that in this pr, hence the suggested name cacheOnlyTopicPool. which now doesn't make much sense

Reply:
> why then they are connected for untill max is discovered? why not to disconnect them immediately when connection was established? so that connection will be used just for validation?

That's a good suggestion. As `MaxReached()` uses cache, I believe that should work just fine. cc @adriacidre

Reply:
@adambabik yes, we can directly drop the discovered mail servers as soon as we confirm the connection can be established. Do you guys also want to change the `cacheOnlyTopicPool` name by something different?

Reply:
Let's try that then. If you have a better name. But I think `cacheOnlyTopicPool` makes sense if we drop the peers immediately after they connect.